rfd/0183-access-request-kube-resource-allow-list.md

```diff
@@ -0,0 +1,306 @@
+
+This is already supported. The admin can use `kubernetes_resources` field to restrict what list of namespaces the requester can list and request for:
+
+```yaml
+kind: role
+metadata:
+  name: kube-access
+spec:
+  allow:
+    # Only lists namespaces starting with "pumpkin-" prefixes
+    kubernetes_resources:
+      - kind: namespace
+        name: pumpkin-*
+        verbs:
+          - list
+    kubernetes_groups:
+      - '{{internal.kubernetes_groups}}'
+    kubernetes_labels:
+      '*': '*'
+    kubernetes_users:
+      - '{{internal.kubernetes_users}}'
+
+----
+
+kind: role
+metadata:
+  name: requester
+spec:
+  allow:
+    request:
+      search_as_roles:
+      - kube-access
+```
+
+Requesting for namespace not in this list is denied.
+
+## As an admin, if a user requests for namespaces, I want to limit namespaces using role templates and traits
+
+Trait interpolation will be supported in [v17](https://github.com/gravitational/teleport/pull/45277) for `kubernetes_resources` field:
+
+```yaml
+kind: user
+metadata:
+  name: lisa@goteleport.com
+spec:
+  created_by:
+    connector:
+      id: okta-integration
+  roles:
+  - requester
+  # Attributes from Okta
+  traits:
+    searchable_kube_namespaces:
+    - pumpkin-*
+
+----
+
+kind: role
+metadata:
+  name: kube-access
+spec:
+  allow:
+    kubernetes_resources:
+      # Only lists namespaces starting with "pumpkin-" prefixes
+      - kind: namespace
+        # Inserts user's trait value
+        name: '{{external.searchable_kube_namespaces}}'
+        verbs:
+          - list
+    kubernetes_groups:
+      - '{{internal.kubernetes_groups}}'
+    kubernetes_labels:
+      '*': '*'
+    kubernetes_users:
+      - '{{internal.kubernetes_users}}
+
+----
+
+kind: role
+metadata:
+  name: requester
+spec:
+  allow:
+    request:
+      search_as_roles:
+      - kube-access
+```
+
+## As an admin, I want to require users to request for Kubernetes subresources instead of the whole Kubernetes cluster
+
+Admins can specify in a role, a Kubernetes request mode option that requires users to select [Kubernetes subresources](https://github.com/gravitational/teleport/blob/110b23aefb3c4b91c9d8cca594041b93f0e078dd/api/types/constants.go#L1233).
+
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  options:
+    request_mode:
+      # Disables requesting whole Kubernetes cluster, but can request any of the kube subresources
+      kubernetes_resources:
+      - kind: *
+```
+
+User requesting kube_cluster will be denied.
+
+## As an admin, I want to limit Kubernetes resource requesting to pods and namespaces only
+
+Admins can specify in a role, a Kubernetes request mode option that requires users to select either pods or namespaces.
+
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  options:
+    request_mode:
+      # Can request only pods or namespaces
+      kubernetes_resources:
+        - kind: namespace
+        - kind: pod
+```
```

Comment:
I'm curious why we have this under `spec.options.request_mode.kubernetes_resources` instead of, say, `spec.allow.request.kubernetes_resources`. Basically why is it a role option and not under `spec.allow.request` where most other access request configuration is? And what is the effect if the option is set differently in multiple roles the user has?

I haven't really thought about it hard enough to say one of those options is definitively better, I'm just wondering if there's a reason or it just seemed like the default way to do it

Reply:
After slightly more thought, I think this doesn't really make sense as a role option, and an allow/deny rule would be better

Here's my mental model of role options (I'm sure there are exceptions to this but i think it's mostly accurate). Usually, role options are coalesced across all roles a user has to result in a single value, and usually that value ends up encoded in the user's certificate. For example, the `forward_agent` option is combined from all roles the user has, and it's encoded in the SSH cert when the user logs in. The same for many other options. Even when not encoded in the cert, usually you compute a single value from the user's entire role set.

If instead you want one role to allow the user to request certain kubernetes namespaces with certain target roles, and another role to allow the user to request kubernetes clusters with a different target role, and another role can deny the user requesting a specific namespace, I think this would be better as an allow/deny rule under `spec.allow.request`

Reply:
@nklaassen updated RFD to reflect field change